File: debian/fop2/plugins/sampleplugin/sampleplugin.pl
# The perl part of a plugin will execute commands on the server side.
# As such, this commands will apply to any user, regardless if the plugin
# is assigned to any user (plugin assignment only affects the client side
# part, or .js files)

# AMI_Event_Handler lets you intercept AMI events and add your own actions/code
# The first parameter is a hash with the key=>value paris as received
#
# The function should return an array with valid AMI cmmands/actions to send
# If you want to see the possible AMI events, you can start fop2_server in debug
# level 1.
#
$AMI_Event_Handler{'sampleplugin'}{'HANGUP'} = sub {
    my $event = shift;
    my @allreturn;

    # Retrieve config data as set in the plugin ini file
    my $var1 = $main::pluginconfig{'sampleplugin'}{'sampleConfig'}{''};
    my $var2 = $main::pluginconfig{'sampleplugin'}{'sampleConfig'}{'samplesection'};

    # This will print out the complete manager event as received
    my @keys =  keys %$event;
    foreach my $key (@keys) {
        print "$key = ".${$event}{$key}."\n";
    }
    print "\n";

    # We can send AMI commands as strings as a reaction to the received AMI event
    $return  = "Action: DBPut\r\n";
    $return .= "Family: SOMETHING\r\n";
    $return .= "Key: TO\r\n";
    $return .= "Val: WRITE\r\n";
    $return .= "\r\n";
    push @allreturn, $return;

    # We return an array containing valid manager Actions
    return @allreturn;
};

# To catch responses from a command sent as a Hash with a special
# formatted ActionID
#
$AMI_Response_Handler{'sampleplugin'}{'myactionresponse'} = sub {

     my $response = shift;
     my $extra    = shift;

     my @allreturn;

     my ($field0, $field1, $field2)   = split/!/, $extra;

     if($response->{Response} eq "Success") {
        # A UserEvent with family and value will be sent to flash clients so we can intercept it in callback_{family} in the plugin js code
        my $valor = { 'Event' => 'UserEvent', 'UserEvent' => 'SomeOtherActionBasedOnResponse', 'Channel' => $field1, 'Family' => "something", 'Value' => $field2 };
        push @allreturn, $valor;

     }

     return @allreturn;
};

# Client_Pre_Command_Handler is called when an action is received from
# a FOP2 Client (from the browser), like changing the presence, initiating
# a transfer, etc. You also can return an AMI command to be sent. 
# The Pre handler will issue your commands BEFORE the standard FOP2
# response. The Post handler will issue the commands AFTER the standard
# FOP2 response.
#
# The first key is the name of the plugin, the second the action received
# from a FOP2 client. You can catch here custom actions defined in the .js
# part of a plugin too.
#
# A good way to see the commands that can be recieved from the client is
# to start fop2_server in debug level 4: /usr/local/fop2/fop2_server -X 4
#
$Client_Pre_Command_Handler{'sampleplugin'}{'ping'} = sub {
    print "pong pong!\n";
    my $pluginconf   = defined($main::pluginconfig{'sampleplugin'}{'sampleconfig'}{'section'})?$main::pluginconfig{'sampleplugin'}{'sampleconfig'}{'section'}:$main::pluginconfig{'sampleplugin'}{'sampleconfig'}{''};
    my @allreturn;

    my $mychannel      = main::get_btn_config( "$contexto", $origen, 'MAINCHANNEL');

    $return  = "Action: DBDel\r\n";
    $return .= "Family: SOMETHING\r\n";
    $return .= "Key: TO\r\n";
    $return .= "\r\n";
    push @allreturn, $return;

    # If we need to get the actual response to our command,
    # then we return a hash instead of a string with a special formaatted ActionID
    # using ! as separator, where the 1st field will be later catched
    # in AMI_Reponse_Handler
    $return = {
        Action    => 'DBPut',
        Family    => 'SOMETHING',
        Key       => 'TO',
        Val       => 'WRITE',
        ActionID  => "myactionresponse!$mychannel!someotherdata"
    };
    push @allreturn, $return;


    push @allreturn, $return;
    return @allreturn;
};

# sampleaction is a custom action sent from the js part of the plugin

$Client_Post_Command_Handler{'sampleplugin'}{'sampleaction'} = sub {
    my @allreturn;
    $return  = "Action: DBDel\r\n";
    $return .= "Family: SOMETHING\r\n";
    $return .= "Key: TO\r\n";
    $return .= "\r\n";
    push @allreturn, $return;
    return @allreturn;
};

File: debian/fop2/plugins/sampleplugin/sampleplugin.js
plugins['sampleplugin'] = (function() {

    // Private variables for the plugin

    // This object can be used as a menu item to be added on the context submenu
    // look at the setExtensionMenu method down bellow. The icon can be set/styled
    // in the plugin .css file. The name  will be used as a method to be called
    // when the menu action is selected.

    var menuOption = {
        "sampleAction": { name: 'Sample Sub Menu Action', icon: 'sampleicon' },
    }

    // Retrieving config data from plugin .ini file 
    var sample1 = parseInt(pluginconfig['sampleplugin']['sampleConfig'][''])
    var sample2 = parseInt(pluginconfig['sampleplugin']['sampleConfig']['samplesection'])

    return { 

        loadLang: function(values) {
            // This method is called to load language files into the application when the plugin is loaded
            // The language files are located in the lang subdirectory of the plugin folder
            //
            // queuecommand is a string that can be used to send a command from the client to the server
            // the xml format can be seen below, it requires to be autehnticated by hashing a key as shown
            // you must pass fields separated by pipes in the format <msg data='position|command|data|hash' />
            //
            // Finally the sendcommand function will send the queuedcommand to the server for processing
            var hash = hex_md5(secret+lastkey);
            queuedcommand = "<msg data=\"" + myposition + "|pluginlang|" + language + "~sampleplugin" + "|" + hash + "\" />";
            sendcommand();
        },

        setLang: function() {
            // This method is used to apply a language strings to html elements
            $('#sampleplugin_element').attr('title',lang.samplepluginstring);
        },

        setExtensionMenu: function(items) {
            // This one is called when the context submenu is opened, it is possible then to 
            // add menu items by extending the default object/arry, in this sample we are adding
            // the private menuOption defined on the top of this file.
            jQuery.extend(items,menuOption);
            return items;
        },
        sampleAction: function(target,source) {
            // this method is called when the custom submenu item is selected. Those menu entries
            // send the target and source extension as the button position in the panel.
            // botonitos is an object that has button properties as defined in buttons.cfg, like
            // extension, channel, label, etc. 
            var number_to_dial = botonitos[target]['EXTENSION'];
            queuedcommand = "<msg data=\"" + myposition + "|dial|" + number_to_dial + "|" + hash + "\" />";
            sendcommand();

        },
        callback_getvar: function(nro,texto,slot) {
             // callback_ methods will intercept events received from the server so you can act upon them
             // These are some of the events that you can receive:
             //
             // voicemailcount,lock,unlock,astdbcust,devtype,fromqueue,queuemembers,settext,usersonline,
             // xstatus,presence,ip,state,link,direction,settimer,rename,notionline,notioffline,details,
             // queueentry,waitingcalls,clearentries,getvar,etc
             //
             // a good way to see the events that are received on the client side is to run fop2_server 
             // with debug level 4: /usr/local/fop2/fop2_server -X 8
             //
             // nro is the button index position
             // texto is the main content
             // slot is the (optional) line number
        },
        init: function() {
            // This method is called on the fop2 initilization, and is used to initialize the plugin itself
            // like for adding page elements, or menu items.
            // embed elements in page, load style files

            if($('#action_sampleAction').length == 0) {
                // add button in action toolbar
                var actionbar = $('#custombar');
                but = document.createElement('div');
                $(but).attr({id: 'action_sampleAction', title: 'Sample'}).addClass('actionbutton myclick');
                img = document.createElement('img');
                $(img).attr('src','./images/toolbar/conference.png');
                $(but).append(img);
                actionbar.append(but);
            }

        }
    }
}());
